File: Components/DevExpressVCL/Demos/ExpressTile Control/CBuilder/RealtorWorld/RealtorWorldSystemInformation.cpp
//---------------------------------------------------------------------------

#include <vcl.h>
#pragma hdrstop

#include "RealtorWorldSystemInformation.h"
//---------------------------------------------------------------------------
#pragma package(smart_init)
#pragma link "RealtorWorldBaseFrame"
#pragma link "cxControls"
#pragma link "cxClasses"
#pragma link "cxLookAndFeels"
#pragma link "cxGraphics"
#pragma link "cxLookAndFeelPainters"
#pragma link "cxEdit"
#pragma link "cxImage"
#pragma link "dxGDIPlusClasses"
#pragma link "cxCustomData"
#pragma link "cxData"
#pragma link "cxDataStorage"
#pragma link "cxFilter"
#pragma link "cxGrid"
#pragma link "cxGridCustomTableView"
#pragma link "cxGridCustomView"
#pragma link "cxGridLevel"
#pragma link "cxGridTableView"
#pragma link "cxGroupBox"
#pragma link "cxLabel"
#pragma link "cxProgressBar"
#pragma link "cxSpinEdit"
#pragma link "cxStyles"

#include "TlHelp32.hpp"
#include "PsAPI.hpp"
#include "Registry.hpp"

#pragma resource "*.dfm"
TfrmSystemInformation *frmSystemInformation;
//---------------------------------------------------------------------------
__fastcall TfrmSystemInformation::TfrmSystemInformation(TComponent* Owner)
	: TfrmBase(Owner)
{
	lbComputerName->Caption = GetMachineName();
	QuerySystemInfo();
    tmRefeshTimer(NULL);
}
//---------------------------------------------------------------------------
String TfrmSystemInformation::GetMachineName()
{
	DWORD ABufferSize;
	char  ABuffer[MAX_PATH + 1];

	ABufferSize = MAX_PATH + 1;
	GetComputerName(&ABuffer[0], &ABufferSize);

	return (String)ABuffer;
}
//---------------------------------------------------------------------------
DWORD TfrmSystemInformation::GetProcessMemoryUsage(THandle AProcessID)
{
	DWORD AResult;
	TProcessMemoryCounters ACounters;
	HANDLE AProcessHandle;

	AResult = 0;
	AProcessHandle = OpenProcess(PROCESS_QUERY_INFORMATION, true, AProcessID);
	if (AProcessHandle)
	{
		ZeroMemory(&ACounters, sizeof(ACounters));
		ACounters.cb = sizeof(ACounters);
		if (GetProcessMemoryInfo(AProcessHandle, &ACounters, sizeof(ACounters)))
			AResult = ACounters.WorkingSetSize;
		CloseHandle(AProcessHandle);
	}
	return AResult;
}
//---------------------------------------------------------------------------
void TfrmSystemInformation::QueryMemoryUsage()
{
	TMemoryStatus AMemStatus;

	ZeroMemory(&AMemStatus, sizeof(AMemStatus));
	AMemStatus.dwLength = sizeof(AMemStatus);
	GlobalMemoryStatus(&AMemStatus);
	pbMemoryUsage->Position = 100.0 * (AMemStatus.dwTotalPhys - AMemStatus.dwAvailPhys) / AMemStatus.dwTotalPhys;
	pbMemoryUsage->Properties->Text = FormatFloat("0.00", AMemStatus.dwAvailPhys / (1024.0 * 1024.0 * 1024.0)) + " GB Free";
}
//---------------------------------------------------------------------------
void TfrmSystemInformation::QuerySystemInfo()
{
	TRegistry *ARegistry;

	ARegistry = new TRegistry();
	ARegistry->RootKey = HKEY_LOCAL_MACHINE;
	if (ARegistry->OpenKeyReadOnly("SOFTWARE\\Microsoft\\Windows NT\\CurrentVersion"))
	{
		lbWindowsInfo->Caption = ARegistry->ReadString("ProductName") + " " + ARegistry->ReadString("CSDVersion");
		ARegistry->CloseKey();
	};
	if (ARegistry->OpenKeyReadOnly("HARDWARE\\DESCRIPTION\\System\\CentralProcessor\\0"))
	{
		lbProcessorInfo->Caption = ARegistry->ReadString("ProcessorNameString");
		ARegistry->CloseKey();
	}
	delete ARegistry;
}
//---------------------------------------------------------------------------
void __fastcall TfrmSystemInformation::tmRefeshTimer(TObject *Sender)
{
	tagPROCESSENTRY32 AProcessInfo;
	int ARecordIndex;
	HANDLE ASnapshot;

	ASnapshot = CreateToolhelp32Snapshot(TH32CS_SNAPPROCESS, 0);
	try
	{
		ZeroMemory(&AProcessInfo, sizeof(AProcessInfo));
		AProcessInfo.dwSize = sizeof(AProcessInfo);
		cxGrid1TableView1->DataController->BeginFullUpdate();
		try
		{
			cxGrid1TableView1->DataController->RecordCount = 0;
			if (Process32First(ASnapshot, &AProcessInfo))
			{
				do
				{
					ARecordIndex = cxGrid1TableView1->DataController->RecordCount;
					cxGrid1TableView1->DataController->RecordCount += 1;
					cxGrid1TableView1->DataController->SetValue(ARecordIndex, cxGrid1TableView1Column1->Index, AProcessInfo.th32ProcessID);
					cxGrid1TableView1->DataController->SetValue(ARecordIndex, cxGrid1TableView1Column2->Index, (String)AProcessInfo.szExeFile);
					cxGrid1TableView1->DataController->SetValue(ARecordIndex, cxGrid1TableView1Column3->Index, GetProcessMemoryUsage(AProcessInfo.th32ProcessID) / 1024);
				}
				while (Process32Next(ASnapshot, &AProcessInfo));
			}
		}
		__finally
		{
			cxGrid1TableView1->DataController->EndFullUpdate();
		}
	}
	__finally
	{
		CloseHandle(ASnapshot);
	}
	QueryMemoryUsage();
}
//---------------------------------------------------------------------------

File: Components/DevExpressVCL/Demos/ExpressTile Control/Delphi/RealtorWorld/RealtorWorldSystemInformation.pas
unit RealtorWorldSystemInformation;

{$I cxVer.inc}

interface

uses
  Windows, Messages, SysUtils, Variants, Classes, Graphics, Controls, Forms, 
  Dialogs, RealtorWorldBaseFrame, dxCustomTileControl, ExtCtrls, cxGraphics,
  cxControls, cxLookAndFeels, cxLookAndFeelPainters, cxStyles, cxCustomData,
  cxFilter, cxData, cxDataStorage, cxEdit, cxGridCustomView,
  cxGridCustomTableView, cxGridTableView, cxClasses, cxGridLevel, cxGrid,
  cxMaskEdit, cxCurrencyEdit, cxSpinEdit, cxContainer, cxGroupBox, cxImage,
  dxGDIPlusClasses, cxLabel, cxProgressBar, dxCore;

type
  TfrmSystemInformation = class(TfrmBase)
    cxGrid1: TcxGrid;
    cxGrid1Level1: TcxGridLevel;
    cxGrid1TableView1: TcxGridTableView;
    cxGrid1TableView1Column1: TcxGridColumn;
    cxGrid1TableView1Column2: TcxGridColumn;
    cxGrid1TableView1Column3: TcxGridColumn;
    tmRefesh: TTimer;
    cxGroupBox1: TcxGroupBox;
    Image1: TcxImage;
    lbComputerName: TcxLabel;
    lbMemory: TcxLabel;
    pbMemoryUsage: TcxProgressBar;
    lbWindowsInfo: TcxLabel;
    lbProcessorInfo: TcxLabel;
    procedure tmRefeshTimer(Sender: TObject);
  private
    function GetMachineName: string;
    function GetProcessMemoryUsage(AProcessID: THandle): Cardinal;
    procedure QueryMemoryUsage;
    procedure QuerySystemInfo;
  public
    constructor Create(AOwner: TComponent); override;
  end;

implementation

uses
  TlHelp32, PsAPI, Registry;

{$R *.dfm}

{ TfrmSystemInformation }

constructor TfrmSystemInformation.Create(AOwner: TComponent);
begin
  inherited Create(AOwner);
  lbComputerName.Caption := GetMachineName;
  QuerySystemInfo;
  QueryMemoryUsage;
  tmRefeshTimer(nil);
end;

function TfrmSystemInformation.GetMachineName: string;
var
  ABuffer: array[0..MAX_PATH + 1] of Char;
  ABufferSize: Cardinal;
begin
  ABufferSize := Length(ABuffer);
  GetComputerName(ABuffer, ABufferSize);
  Result := ABuffer;
end;

function TfrmSystemInformation.GetProcessMemoryUsage(AProcessID: THandle): Cardinal;
var
  ACounters: TProcessMemoryCounters;
  AProcessHandle: THandle;
begin
  Result := 0;

  AProcessHandle := OpenProcess(PROCESS_QUERY_INFORMATION, True, AProcessID);
  if AProcessHandle <> 0 then
  try
    ZeroMemory(@ACounters, SizeOf(ACounters));
    ACounters.cb := SizeOf(ACounters);
    if GetProcessMemoryInfo(AProcessHandle, @ACounters, SizeOf(ACounters)) then
      Result := ACounters.WorkingSetSize;
  finally
    CloseHandle(AProcessHandle);
  end;
end;

procedure TfrmSystemInformation.tmRefeshTimer(Sender: TObject);
var
  AProcessInfo: TProcessEntry32;
  ARecordIndex: Integer;
  ASnapshot: THandle;
begin
  ASnapshot := CreateToolhelp32Snapshot(TH32CS_SNAPPROCESS, 0);
  try
    ZeroMemory(@AProcessInfo, SizeOf(AProcessInfo));
    AProcessInfo.dwSize := SizeOf(AProcessInfo);
    cxGrid1TableView1.DataController.BeginFullUpdate;
    try
      cxGrid1TableView1.DataController.RecordCount := 0;
      if Process32First(ASnapshot, AProcessInfo) then
      repeat
        ARecordIndex := cxGrid1TableView1.DataController.RecordCount;
        cxGrid1TableView1.DataController.RecordCount := cxGrid1TableView1.DataController.RecordCount + 1;
        cxGrid1TableView1.DataController.Values[ARecordIndex, cxGrid1TableView1Column1.Index] :=
          AProcessInfo.th32ProcessID;
        cxGrid1TableView1.DataController.Values[ARecordIndex, cxGrid1TableView1Column2.Index] :=
          dxStringToAnsiString(AProcessInfo.szExeFile);
        cxGrid1TableView1.DataController.Values[ARecordIndex, cxGrid1TableView1Column3.Index] :=
          GetProcessMemoryUsage(AProcessInfo.th32ProcessID) div 1024;

      until not Process32Next(ASnapshot, AProcessInfo);
    finally
      cxGrid1TableView1.DataController.EndFullUpdate;
    end;
    QueryMemoryUsage;
  finally
    CloseHandle(ASnapshot);
  end;
end;

procedure TfrmSystemInformation.QueryMemoryUsage;
const
  GigaByte = 1024 * 1024 * 1024;
var
{$IFDEF DELPHI12}
  AMemStatus: TMemoryStatusEx;
{$ELSE}
  AMemStatus: TMemoryStatus;
{$ENDIF}
begin
  ZeroMemory(@AMemStatus, SizeOf(AMemStatus));
  AMemStatus.dwLength := SizeOf(AMemStatus);
{$IFDEF DELPHI12}
  if GlobalMemoryStatusEx(AMemStatus) then
  begin
    pbMemoryUsage.Position := 100 * (AMemStatus.ullTotalPhys - AMemStatus.ullAvailPhys) / AMemStatus.ullTotalPhys;
    pbMemoryUsage.Properties.Text := Format('%0.2f GB Free', [AMemStatus.ullAvailPhys / GigaByte]);
  end;
{$ELSE}
  GlobalMemoryStatus(AMemStatus);
  pbMemoryUsage.Position := 100 * (AMemStatus.dwTotalPhys - AMemStatus.dwAvailPhys) / AMemStatus.dwTotalPhys;
  pbMemoryUsage.Properties.Text := Format('%0.2f GB Free', [AMemStatus.dwAvailPhys / GigaByte]);
{$ENDIF}
end;

procedure TfrmSystemInformation.QuerySystemInfo;
var
  ARegistry: TRegistry;
begin
  ARegistry := TRegistry.Create;
  try
    ARegistry.RootKey := HKEY_LOCAL_MACHINE;
    if ARegistry.OpenKeyReadOnly('SOFTWARE\Microsoft\Windows NT\CurrentVersion') then
    try
      lbWindowsInfo.Caption := ARegistry.ReadString('ProductName') + ' ' + ARegistry.ReadString('CSDVersion');
    finally
      ARegistry.CloseKey;
    end;
    if ARegistry.OpenKeyReadOnly('HARDWARE\DESCRIPTION\System\CentralProcessor\0') then
    try
      lbProcessorInfo.Caption := ARegistry.ReadString('ProcessorNameString');
    finally
      ARegistry.CloseKey;
    end;
  finally
    ARegistry.Free;
  end;
end;

initialization
  RegisterFrame(IDSystemInformation, TfrmSystemInformation);
end.
